sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusMessageBatch.cs

```diff
@@ -74,14 +77,42 @@ internal ServiceBusMessageBatch(TransportMessageBatch transportBatch)
         ///
         public bool TryAdd(ServiceBusMessage message)
         {
+            if (_locked)
+            {
+                throw new InvalidOperationException(Resources.MessageBatchIsLocked);
+            }
+
             return InnerBatch.TryAdd(message);
         }
 
         /// <summary>
         ///   Performs the task needed to clean up resources used by the <see cref="ServiceBusMessageBatch" />.
         /// </summary>
         ///
-        public void Dispose() => InnerBatch.Dispose();
+        public void Dispose()
+        {
+            if (_locked)
```

Comment:
nit: consider adding ValidateNotLocked method.

Reply:
Because we're using this as a `volatile` field and we have an (acceptably) small race condition, introducing a method call here wouldn't likely be the best path unless we marked it up for aggressive inlining.     

That said, you made me take a closer look and realize that we really should have the full set of synchronization concerns in the outer batch and leave the transport batch as-is.   That let us drop the `volatile` semantics and just use the same sync root across operations.   So... you got your method.  :smile: